engine/src/main/java/org/pentaho/di/trans/steps/script/ScriptAddedFunctions.java

```diff
@@ -0,0 +1,2532 @@
+          Calendar startDate = Calendar.getInstance();
+          startDate.setTime( dIn );
+          if ( strType.equals( "y" ) ) {
+            return new Double( startDate.get( Calendar.DAY_OF_YEAR ) );
+          } else if ( strType.equals( "m" ) ) {
+            return new Double( startDate.get( Calendar.DAY_OF_MONTH ) );
+          } else if ( strType.equals( "w" ) ) {
+            return new Double( startDate.get( Calendar.DAY_OF_WEEK ) );
+          } else if ( strType.equals( "wm" ) ) {
+            return new Double( startDate.get( Calendar.DAY_OF_WEEK_IN_MONTH ) );
+          }
+          return new Double( startDate.get( Calendar.DAY_OF_YEAR ) );
+        }
+      } catch ( Exception e ) {
+        return null;
+        // throw new RuntimeException(e.toString());
+      }
+    } else {
+      throw new RuntimeException( "The function call getDayNumber requires 2 arguments." );
+    }
+  }
+
+  public static Object isWorkingDay( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+    if ( ArgList.length == 1 ) {
+      try {
+        if ( isNull( ArgList[0] ) ) {
+          return null;
+        } else if ( isUndefined( ArgList[0] ) ) {
+          return undefinedValue;
+        } else {
+          java.util.Date dIn = (java.util.Date) ArgList[0];
+          Calendar startDate = Calendar.getInstance();
+          startDate.setTime( dIn );
+          if ( startDate.get( Calendar.DAY_OF_WEEK ) != Calendar.SATURDAY
+            && startDate.get( Calendar.DAY_OF_WEEK ) != Calendar.SUNDAY ) {
+            return Boolean.TRUE;
+          }
+          return Boolean.FALSE;
+        }
+      } catch ( Exception e ) {
+        return null;
+      }
+    } else {
+      throw new RuntimeException( "The function call isWorkingDay requires 1 argument." );
+    }
+  }
+
+  @SuppressWarnings( "unused" )
+  public static Object fireToDB( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+
+    Object oRC = new Object();
+    if ( ArgList.length == 2 ) {
+      try {
+        Object scmO = actualObject.get( "_step_" );
+        Script scm = (Script) scmO;
+        String strDBName = (String) ArgList[0];
+        String strSQL = (String) ArgList[1];
+        DatabaseMeta ci = DatabaseMeta.findDatabase( scm.getTransMeta().getDatabases(), strDBName );
+        if ( ci == null ) {
+          throw new RuntimeException( "Database connection not found: " + strDBName );
+        }
+        ci.shareVariablesWith( scm );
+
+        Database db = new Database( scm, ci );
+        db.setQueryLimit( 0 );
+        try {
+          if ( scm.getTransMeta().isUsingUniqueConnections() ) {
+            synchronized ( scm.getTrans() ) {
+              db.connect( scm.getTrans().getTransactionId(), scm.getPartitionID() );
+            }
+          } else {
+            db.connect( scm.getPartitionID() );
+          }
+
+          ResultSet rs = db.openQuery( strSQL );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *rs*. The resource is referenced at lines 502, 506, 509, 512, 513, 515, 517, 520. The resource is closed at line 526. There are other execution paths that do not contain closure statements, for example, when ResultSet.next throws an exception. Ensure *rs* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).